### PR TITLE
Add caching in ComponentBase

### LINKED ISSUE #2505: [feature_request] Cachable component
Hi OctoberCMS team,

I would like to have a feature that allows me to configure the caching behavior of a partial view in a component. To support that feature, I would like to add this class into the core of OctoberCMS (if you don't mind).

Please check my code at: https://gist.github.com/khoatran/f6fadc3a70b577910fc55408d66ed8a6 for the CachableComponent class and the example of how to use it.

Cachable components/widget actually is a very popular concept in other CMS. So, it's great if OctoberCMS can support this.

Best regards,

### PR DESCRIPTION
**Pull request by Samuell1**
Resolves #2505

### CODE REVIEW

File: modules/cms/classes/ComponentBase.php [hunk 1]
```diff
@@ -38,6 +39,11 @@ abstract class ComponentBase extends Extendable
      */
     public $isHidden = false;
 
+    /**
+     * @var int Determines if component should cache default partial in onRender hook. Uses minutes.
```
Comment (LukeTowers) [inline modules/cms/classes/ComponentBase.php]: This should be in seconds, that's what Laravel 5.9 is moving to so might as well do it here.

File: modules/cms/classes/ComponentBase.php [hunk 2]
```diff
@@ -131,6 +137,11 @@ public function onRun()
      */
     public function onRender()
     {
+        if ($this->cacheTTL) {
+            return Cache::remember($this->getCacheKey(), now()->addMinutes($this->cacheTTL), function () {
```
Comment (LukeTowers) [inline modules/cms/classes/ComponentBase.php]: cacheTTL should also support being a datetime value in which case it should just be passed straight to Cache::remember

File: modules/cms/classes/ComponentBase.php [hunk 3]
```diff
@@ -326,4 +337,12 @@ public function __toString()
     {
         return $this->alias;
     }
+
+    /**
+     * Returns the component's cache key
+     */
+    public function getCacheKey()
+    {
+        return 'component::' . $this->dirName;
```
Comment (LukeTowers) [inline modules/cms/classes/ComponentBase.php]: This should also have the name and alias of the component in here to be as unique as possible

File: modules/cms/classes/ComponentBase.php [hunk 4]
```diff
@@ -40,7 +40,7 @@ abstract class ComponentBase extends Extendable
     public $isHidden = false;
 
     /**
-     * @var int Determines if component should cache default partial in onRender hook. Uses minutes.
+     * @var int Determines if component should cache default partial in onRender hook. Supports seconds or datetime.
```
Comment (LukeTowers) [inline modules/cms/classes/ComponentBase.php]: Typehint should also be updated to `integer|DateTime` please

File: modules/cms/classes/ComponentBase.php [hunk 5]
```diff
@@ -138,7 +138,7 @@ public function onRun()
     public function onRender()
     {
         if ($this->cacheTTL) {
-            return Cache::remember($this->getCacheKey(), now()->addMinutes($this->cacheTTL), function () {
+            return Cache::remember($this->getCacheKey(), $this->cacheTTL, function () {
```
Comment (LukeTowers) [inline modules/cms/classes/ComponentBase.php]: You'll need to do it like the following to support the pre and post Laravel 6 upgrade:

```php
$expires = is_integer($this->cacheTTL) ? now()->addSeconds($this->cacheTTL) : $this->cacheTTL;
```

Comment (LukeTowers) [reply modules/cms/classes/ComponentBase.php]: and then use $expires in Cache::remember

File: modules/cms/classes/ComponentBase.php [hunk 6]
```diff
@@ -326,4 +337,12 @@ public function __toString()
     {
         return $this->alias;
     }
+
+    /**
+     * Returns the component's cache key
+     */
+    public function getCacheKey()
+    {
+        return 'component::' . $this->dirName . $this->alias . $this->name;
```
Comment (LukeTowers) [inline modules/cms/classes/ComponentBase.php]:
```suggestion
        return 'component::' . $this->dirName . '-' . $this->page->getFilePath() . '-' . $this->alias;
```

**Comment by bennothommo** [comment]
@nathan-van-der-werf @prhost Any particular reason for the thumbs down on this PR?

**Comment by prhost** [comment]
> @nathan-van-der-werf @prhost Any particular reason for the thumbs down on this PR?

If an extra dependency is being created in a place it should not have, this problem can easily be solved by placing it in the component itself, or even using events to inject this cache system, outside of the developers' own custom, including the actual one. October documentation, that onRender doesn't call parent, then no immediate component would take effect, and if you have to change all or most of the onRender and call parent, it's easier to implement this snippet by hand when you need it.

The idea is cool, but I suggest @Samuell1 checking if there are any events that can inject this and create a plugin.

**Comment by Samuell1** [comment]
Yeah, i was talking with @LukeTowers few days back that most of the components are dynamic that means they cant cache or they have custom onRender method to handle things and that means caching in components base is useless.

Its not my idea its from issue https://github.com/octobercms/october/issues/2505 @prhost

**Pull request closed without merging**